RELEASE/scripts/autoscend/auto_quest_level_8.ash

```diff
@@ -13,54 +13,263 @@ boolean L8_trapperStart()
+		// mine the square directly above it
+		return lastCell - 8;
+	}
+
+	// - If we've got to row 3 or above, start searching for ores.
+	item[int] minedCells = parseMineLayout();
+	int[int] oreSeen;
+	foreach oreCell, oreType in minedCells {
+		if (oreType == oreGoal) {
+			oreSeen[oreCell] = 1; // value doesn't matter, just want to count and iterate the keys
+		}
+	}
+	int[int] sparklingCells = findSparklingCells(mineLayout);
+	int[int] potentialCells;
+	int potentialCount = 0;
+	if (count(oreSeen) == 0) {
+		// - Not found any ore that we're looking for yet
+		if (lastCell > 24 && lastCell < 31) {
+			// get to row 2 as our probability of hitting ore we're looking for is higher.
+			return lastCell - 8;
 		}
-		autoAdv(1, $location[The Goatlet]);
-		auto_sourceTerminalEducate($skill[Extract], $skill[Portscan]);
+		// find all the sparkling tiles in the top n rows
+		// start from the top 2, if we haven't found any there,
+		// increase the search space by 1 row and check again until we max out at the 4th row
+		// avoid columns 1 and 6 as they limit the search space since we can't mine column 0 or 7.
+		// unless we run into a situation where we've mined all the other sparkling cells.
+		int rowLimit = 2;
+		boolean avoidSides = true;
+		while (count(potentialCells) == 0 && rowLimit < 5) {
+			foreach sparkleCell in sparklingCells {
+				if (canMine(sparkleCell, rowLimit)) {
+					if (!isInSideColumn(sparkleCell) || !avoidSides) {
+						potentialCells[potentialCount] = sparkleCell;
+						potentialCount++;
+					}
+				}
+			}
+			rowLimit++;
+			if (avoidSides && rowLimit == 5 && count(potentialCells) == 0) {
+				avoidSides = false;
+				rowLimit = 2;
+			}
+		}
+	} else {
+		// - Found at least one ore that we're looking for!
+		// search orthogonally from the cells we found our required ore in as ore is always contiguous
+		// limit our search to the top 3 rows to begin, if we don't find any cells that meet the criteria
+		// increase the limit to the top 4 rows and check again.
+		int rowLimit = 3;
+		while (count(potentialCells) == 0 && rowLimit < 5) {
+			foreach oreCell in oreSeen {
+				int[4] orthogonals = getOrthogonals(oreCell);
+				foreach orthoPos, orthoCell in orthogonals {
+					if (canMine(orthoCell, rowLimit) && sparklingCells contains orthoCell) {
+						potentialCells[potentialCount] = orthoCell;
+						potentialCount++;
+					}
+				}
+			}
+			rowLimit++;
+		}
+	}
+	// only found one potential, just return it
+	if (count(potentialCells) == 1) {
+		return potentialCells[0];
+	}
+	// found 2 or more potentials, return a random one of them
+	return potentialCells[random(count(potentialCells))];
+}
+
+void auto_testMining(item oreGoal) {
+	// use this to test in aftercore. Will mine 3 ores of whatever type you pass e.g.
+	// ash import <autoscend.ash> auto_testMining($item[linoleum ore]);
+	// I'd recommend equipping some HP regen stuff in slots not needed by the Mining Gear
```

Comment:
why not have it automatically equip HP regen gear in slots not taken up by the outfit?

Reply:
Because you're supposed to be in aftercore and this is a test function. I really didn't want to make it overly complicated and have it fail for something unrelated to testing the functionality it is supposed to test.

Also the amount of HP you lose mining here is trivial even in run so it's unlikely it'll be a problem. I just put the information there incase someone has a problem. When `auto_mineForOres` gets removed and this is no longer opt-in, this function will be removed too.

Reply:
Sorry, let me clarify.
Why not put the same code snippet for maximizing HP recovery in both the in-run and test function. with the latter only being there to test its functionality for the former.

Reply:
Actually I probably should put a call to `equipMaximizedGear()` in `L8_getMineOres()` before equipping the Mining Gear to take care of that in-run.